### PR TITLE
First draft

### PR DESCRIPTION
**Pull request by amexboy**

### CODE REVIEW

File: src/github/amexboy/core.clj [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      (create-project  source-root name branch modules options)
+      (println "Something went wrong. Missing either branch or modules. -h for help" options))
+    ))
+
+(defn -main [& args]
+  (let [{:keys [options errors summary]} (opts/parse-opts args cli-options :strict true)]
+    (cond
+      (or errors (:help options)) (println errors "\n Usage: \n" summary)
+      (validate-options options) (dispatch options)
+      :default
```
Comment (RadicalZephyr) [inline src/github/amexboy/core.clj]: You don't need the `:default` case since there's nothing to return here.

File: src/github/amexboy/core.clj [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      ))
+  )
```
Comment (RadicalZephyr) [inline src/github/amexboy/core.clj:47]: Breaking ending parens on their own lines is considered bad style, very little clojure (or other lisp code) in the wild is formatted like this.

File: src/github/amexboy/git_ops.clj [hunk 3]
```diff
@@ -0,0 +1,37 @@
+(ns github.amexboy.git-ops
+  (:use [github.amexboy.shell :only [sh]]))
```
Comment (RadicalZephyr) [inline src/github/amexboy/git_ops.clj:2]: `:use` is not recommended anymore, the preferred way to do this is with `:require` and `:refer` like `:require [github.amexboy.shell :refer [sh]`

File: src/github/amexboy/modules_loader.clj [hunk 4]
```diff
@@ -0,0 +1,52 @@
+(ns github.amexboy.modules-loader
+  (:require [clojure.java.io :as io])
+  (:use [clojure.java.shell :only [sh]]
+        [github.amexboy.git-ops]
+        [github.amexboy.project-gen]))
```
Comment (RadicalZephyr) [inline src/github/amexboy/modules_loader.clj]: It's not really a good practice to refer whole namespaces, especially multiple in one namespace because it makes it really hard to find where functions come from.  `:refer [... :as ...]` is a much better idea.  It works fine on a small project but it really doesn't scale well.  This is more of a mild suggestion, I totally do this sometimes too when I'm feeling lazy 😉

File: src/github/amexboy/project_gen.clj [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+(defn maven-root? [dir]
+  "Checks if the dir contains a maven.pom file"
+  (some (partial = "pom.xml") (.list dir))
+  )
+(defn gradle-root? [dir]
+  "Checks if the dir contains a maven.pom file"
+  (some (partial = "build.gradle") (.list dir))
+  )
+(defn liengen-root? [dir]
```
Comment (RadicalZephyr) [inline src/github/amexboy/project_gen.clj:13]: misspelled leiningen 😉

File: src/github/amexboy/project_gen.clj [hunk 6]
```diff
@@ -0,0 +1,136 @@
+(ns github.amexboy.project-gen
+  (:require [clojure.string :as st]
+            [clojure.java.io :as io]))
+
+(defn maven-root? [dir]
+  "Checks if the dir contains a maven.pom file"
+  (some (partial = "pom.xml") (.list dir))
```
Comment (RadicalZephyr) [inline src/github/amexboy/project_gen.clj:7]: You could make a higher-order function to make all three of these conditionals, call it something like `dir-contains-file?`  I mainly suggest this because of the use of Java interop and because I'm not sure if you're familiar with the technique, it's a good one to have in your toolkit.

File: src/github/amexboy/project_gen.clj [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  <component name=\"GradleSettings\">
+    <option name=\"linkedExternalProjectsSettings\">
+      <GradleProjectSettings>
+        <option name=\"distributionType\" value=\"DEFAULT_WRAPPED\" />
+        <option name=\"externalProjectPath\" value=\"$PROJECT_DIR$/" module "/\" />
+        <option name=\"useQualifiedModuleNames\" value=\"true\" />
+      </GradleProjectSettings>
+    </option>
+ </component>
+  ")
```
Comment (RadicalZephyr) [inline src/github/amexboy/project_gen.clj:59]: For large string interpolation like this I would use `format` instead of `str` because it makes it MUCH easier to see what the interpolated values are, especially since these strings have so many `"` in them.

File: src/github/amexboy/project_gen.clj [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+(defn gen-liengen-iml [path name]
+  "Generates an IDEA module for the path with the given name"
+  (let [
+        template "<?xml version=\"1.0\" encoding=\"UTF-8\"?>
+                     <module cursive.leiningen.project.LeiningenProjectsManager.displayName=\"" name "\"
+                             cursive.leiningen.project.LeiningenProjectsManager.isLeinModule=\"true\"
+                             type=\"JAVA_MODULE\" version=\"4\">
+                     <component name=\"ExternalSystem\" externalSystem=\"Maven\" />
+                  </module>"]
```
Comment (RadicalZephyr) [inline src/github/amexboy/project_gen.clj:44]: It looks like this is supposed to interpolate `name` into this string but that's not actually happening right now. As my other comment says, I would use `format` to do that.

File: src/github/amexboy/project_gen.clj [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+ </component>
+  ")
+  )
+
+(defn module-entries [root module]
+  "Returns component entries for git, pom (gradle soon) and git component lists"
+  (let [entries {
+                 :iml (str "<module filepath=\"file://$PROJECT_DIR$/" module "/" module ".iml\"")
+                 :git (str "<mapping directory=\"$PROJECT_DIR$/" module "\" vcs=\"Git\" />")
+                 }
```
Comment (RadicalZephyr) [inline src/github/amexboy/project_gen.clj:67]: This is the canonical way to format this map:

``` clojure
(let [entries {:iml (str "<module filepath=\"file://$PROJECT_DIR$/" module "/" module ".iml\"")
               :git (str "<mapping directory=\"$PROJECT_DIR$/" module "\" vcs=\"Git\" />")}
    ...)
```

File: src/github/amexboy/project_gen.clj [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+        module-file (io/file root module)]
+    (cond
+      (maven-root? module-file) (assoc entries :pom (str "<option value=\"$PROJECT_DIR$/" module "/pom.xml\" />"))
+      (gradle-root? module-file) (assoc entries :gradle (gradle-entry module))
+      ))
+  )
+
+(defn reducer [r v]
+  (reduce #(assoc %1 %2 (cons (%2 r) (%2 v)))
+          r (keys v)))
```
Comment (RadicalZephyr) [inline src/github/amexboy/project_gen.clj:77]: This is pretty dense to use the simple anonymous function syntax on, I would use `fn` and give the variable names and use longer names for `r` and `v`.

This can also be re-written more simply I think.  It looks like this is just taking two maps that are expected to have the same set of keys and merging their values by concatenating them, so you could write this as:

`(merge cons r v)`

Also, I think you want to use `conj` instead.  In general, I never use `cons` because `conj` does exactly the same thing when given a list, but is more general and idiomatic.  Also, since the argument order is more consistent that is also nice.

File: src/github/amexboy/project_gen.clj [hunk 11]
```diff
@@ -0,0 +1,129 @@
+(ns github.amexboy.project-gen
+  (:require [clojure.string :as st]
```
Comment (RadicalZephyr) [inline src/github/amexboy/project_gen.clj:2]: `clojure.string` is _usually_ import as `str`, it won't interfere with the `str` function.  Just a suggestion tho, not a requirement.

**Review by RadicalZephyr — COMMENTED** [review]
You're doing a lot of things right, keep it up!  Hope my comments are helpful.  I notice that a few of my comments were made on reddit as well so feel free to ignore those.

**Comment by amexboy** [comment]
> You're doing a lot of things right, keep it up! Hope my comments are helpful. I notice that a few of my comments were made on reddit as well so feel free to ignore those.

@RadicalZephyr I love the comments, I'll remember all of them. Thank you very much for taking the time!  (I'm now a follower of your blog)

File: src/github/amexboy/core.clj [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+   ["-v" nil "Verbosity level"
+    :id :verbosity
+    :default 0
+    :update-fn inc]                                         ; Prior to 0.4.1, you would have to use:
+   ;; :assoc-fn (fn [m k _] (update-in m [k] inc))
+   ;; A boolean option defaulting to nil
+   ["-h" "--help"]])
+
+(defn validate-options [options]
+  (every? #(options %) [:branch :modules]))
```
Comment (potetm) [inline src/github/amexboy/core.clj:32]: I would use `contains?` here. It's unclear that `options` is a hash map because it's passed in:

```clj
(every? #(contains? options %)
        [:branch :modules])
```

File: src/github/amexboy/core.clj [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  (if list
+    (list-modules source-root)
+    (create-project  source-root name branch modules options)))
+
+(defn -main [& args]
+  (let [{:keys [options errors summary]} (opts/parse-opts args cli-options :strict true)]
+    (cond
+      (or errors (:help options)) (println errors "\n Usage: \n" summary)
+      (validate-options options) (dispatch options)
+      :default (println "Missing either branch or modules\n Usage: \n" summary)
```
Comment (potetm) [inline src/github/amexboy/core.clj:45]: `:default` is for multi-methods. It's normal to use `:else` as a fallthrough for `cond` expressions.

File: src/github/amexboy/modules_loader.clj [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+        work-tree-file (io/file root (:modules-root opts) name module)
+        work-tree (.getPath work-tree-file)]
+    (if (update-repository root module opts)
+      (do
+        (println module)
+        (git-add-worktree src branch work-tree (:base-branch opts))
+        (cond
+          (maven-root? work-tree-file) (gen-maven-iml work-tree module)
+          (gradle-root? work-tree-file) (gen-gradle-iml work-tree module)
+          (liengen-root? work-tree-file) (gen-liengen-iml work-tree module)
```
Comment (potetm) [inline src/github/amexboy/modules_loader.clj:52]: `leiningen`

File: src/github/amexboy/modules_loader.clj [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+          )
+        module)
+      false)))
+
+(defn create-project
+  "Creates a new ipr file and loads the module checking out the branch on the modules"
+  [root name branch modules opts]
+  (gen-ipr (io/file root (:modules-root opts) name)
+           name
+           (filter identity (map #(create-module root % branch name opts) modules))
```
Comment (potetm) [inline src/github/amexboy/modules_loader.clj:62]: Use `keep` instead of `(filter identity ...)`. `identity` is almost never useful in production code, though it's very useful at the REPL.

Comment (potetm) [reply src/github/amexboy/modules_loader.clj:62]: `create-module` appears to be writing to disk. Don't do side-effects inside sequence functions (e.g. `map`, `filter`). Do a `doseq`, and _then_ continue other work:

```clj
(doseq [m module]
  (create-module ... m ...))
(gen-ipr ...)
```

If you need to track some state, I recommend using `loop/recur` instead:

```clj
(loop [mods modules
       created #{}]
  (when-some [m (first mods)]
    (if (create-module ...)
      (recur (next mods) (conj created m)
      (recur (next mods) created)))
```

Though, I think that if you iron all of this disparate logic into larger chunks, you'll find you don't need to track quite so much state in the first place. So I would be hesitant to jump right to `loop/recur`.

**Comment by potetm** [comment]
Some general feedback:

I think you've over-carved this problem. You have 5 namespaces and dozens of functions and only a couple-hundred lines of code. It makes it _very_ hard to read at a glance because you must jump around between functions and namespaces, and you must be aware of what each parameter to a function means. Many of your functions have 3, 4, or even 5 parameters. This indicates to me that there is likely a lot of unnecessary passing of information back and forth for no particular benefit.

The best advice I can give is to write _everything_ you want to do in one long function. That way you can see it all laid out.

Once you've done that, my rule is: a function (or namespace) must justify its existence. If two code paths need to run through the same routine, you _must_ create a function. If, after you write everything out, it's difficult to follow, then and _only then_ do you create named "helper functions." These "helper functions" should start their life inside a `let` statement to make it clear the scope of their utility. If you find yourself often wanting to run the helpers in isolation, then and only then do you move them to the namespace scope.

The word "module" appears to be _heavily_ overloaded in this system. It's _very_ difficult to tell what you're referring to for any given usage. For example, the `create-module` fn accepts `module` as a parameter, and, presumably, creates some other module.* I think this is, in part, a product of over-carving the problem. You end up having to create names for things that are fairly ambiguous and transitory.

Lastly, I want to second some of the things @RadicalZephyr said. You should avoid `%1 %2`. If an anonymous fn takes multiple args, just use `fn`. Beware `cons`. I've literally never used it in years of production work. Don't use `:use`. It makes things much hard to follow in order to save a couple of characters per call-site. I don't ever use `:refer` either, for the same reason.

---

\* On second read, `create-module` appears to be side-effecty. I noted it in another comment, but your usage didn't make that clear at all. Perhaps "module" usually means "module-name"?

**Comment by amexboy** [comment]
@potetm Thank you so so much. I learned so much from your comments

File: src/github/amexboy/git_ops.clj [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+
+(defn git-prune-worktree [root]
+  "Removes dead worktrees"
```
Comment (theodesp) [inline src/github/amexboy/git_ops.clj:5]: you may want to combine them in a defmulti method for all git operations
See http://clojuredocs.org/clojure.core/defmulti

File: src/github/amexboy/modules_loader.clj (hunk 15 above)
Comment (lsh-0) [reply src/github/amexboy/modules_loader.clj:62]: > Use 'keep' instead

huh - this is new trick for me. I've been using `(remove nil? (...))` for a long time. `(filter identity (...))` would have never occurred to me

File: src/github/amexboy/project_gen.clj [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+(defn gen-gradle-iml [path name]
+  "Generates an IDEA module for the path with the given name"
+  (let [template "<?xml version=\"1.0\" encoding=\"UTF-8\"?>
+           <module org.jetbrains.idea.gradle.project.GradlenProjectsManager.isGradleModule=\"true\" type=\"JAVA_MODULE\" version=\"4\">
+           <component name=\"ExternalSystem\" externalSystem=\"Maven\" />
+        </module>"]
+    (spit (str path "/" name ".iml") template))
+  )
+
+(defn gen-liengen-iml [path name]
```
Comment (lsh-0) [inline src/github/amexboy/project_gen.clj:36]: another misspelling of 'leiningen'

File: src/github/amexboy/git_ops.clj (hunk 3 above)
Comment (nbotelhodev) [reply src/github/amexboy/git_ops.clj:2]: You also can use `:require [github.amexboy.shell :as sh]`
and call something like `(sh/ls dir)`
